File: 23-12-2024_Minimum_Number_of_Operations_to_Sort_a_Binary_Tree_by_Level.py
from collections import deque

class Solution:

    # Helper function to calculate the minimum number of swaps to sort the array
    def minSwaps(self, arr):
        n = len(arr)
        
        # Create a list of pairs where each pair is (value, original index)
        v = [(val, idx) for idx, val in enumerate(arr)]
        
        # Sort the list based on values
        v.sort(key=lambda x: x[0])
        
        visited = [False] * n
        swaps = 0
        
        # Iterate over the sorted list to calculate minimum swaps
        for i in range(n):
            # If already visited or already in the correct place, continue
            if visited[i] or v[i][1] == i:
                continue
            
            # Start the cycle of swapping elements
            cycle_size = 0
            j = i
            while not visited[j]:
                visited[j] = True
                j = v[j][1]  # Move to the next element in the cycle
                cycle_size += 1
            
            # If there is a cycle of size n, we need (n-1) swaps to sort it
            if cycle_size > 1:
                swaps += (cycle_size - 1)
        
        return swaps

    # Function to compute the minimum operations on the binary tree
    def minimumOperations(self, root):
        if not root:
            return 0
        
        # Queue for level-order traversal
        q = deque([root])
        ans = 0
        
        # Perform level-order traversal
        while q:
            sz = len(q)
            temp = []
            
            # Process each level
            for _ in range(sz):
                curr_node = q.popleft()
                temp.append(curr_node.val)
                
                # Add child nodes to the queue
                if curr_node.left:
                    q.append(curr_node.left)
                if curr_node.right:
                    q.append(curr_node.right)
            
            # For this level, calculate the minimum swaps required
            ans += self.minSwaps(temp)
        
        return ans


# Definition for a binary tree node
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

File: 23-12-2024_Minimum_Number_of_Operations_to_Sort_a_Binary_Tree_by_Level.java
import java.util.*;

public class Solution {

    // Helper function to calculate the minimum number of swaps to sort the array
    public int minSwaps(List<Integer> arr) {
        int n = arr.size();

        // Create a list of pairs where each pair is {value, original index}
        List<Map.Entry<Integer, Integer>> v = new ArrayList<>();
        for (int i = 0; i < n; ++i) {
            v.add(new AbstractMap.SimpleEntry<>(arr.get(i), i));
        }

        // Sort the list based on values
        Collections.sort(v, (a, b) -> Integer.compare(a.getKey(), b.getKey()));

        boolean[] visited = new boolean[n];
        int swaps = 0;

        // Iterate over the sorted list to calculate minimum swaps
        for (int i = 0; i < n; ++i) {
            // If already visited or already in the correct place, continue
            if (visited[i] || v.get(i).getValue() == i) {
                continue;
            }

            // Start the cycle of swapping elements
            int cycleSize = 0;
            int j = i;
            while (!visited[j]) {
                visited[j] = true;
                j = v.get(j).getValue(); // Move to the next element in the cycle
                cycleSize++;
            }

            // If there is a cycle of size n, we need (n-1) swaps to sort it
            if (cycleSize > 1) {
                swaps += (cycleSize - 1);
            }
        }

        return swaps;
    }

    // Function to compute the minimum operations on the binary tree
    public int minimumOperations(TreeNode root) {
        if (root == null) {
            return 0;
        }

        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        int ans = 0;

        // Perform level-order traversal
        while (!q.isEmpty()) {
            int sz = q.size();
            List<Integer> temp = new ArrayList<>();

            // Process each level
            for (int i = 0; i < sz; i++) {
                TreeNode currNode = q.poll();
                temp.add(currNode.val);

                // Push the child nodes to the queue for next level
                if (currNode.left != null) {
                    q.add(currNode.left);
                }
                if (currNode.right != null) {
                    q.add(currNode.right);
                }
            }

            // For this level, calculate the minimum swaps required
            ans += minSwaps(temp);
        }

        return ans;
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
    }
